runner/app/live/streamer/process_guardian.py

```diff
@@ -0,0 +1,276 @@
+import asyncio
+import logging
+import time
+from typing import Awaitable, Callable, Optional
+from trickle import InputFrame, OutputFrame
+
+from .process import PipelineProcess
+from .status import PipelineState, PipelineStatus
+
+
+class ProcessGuardian:
+    """
+    This class is responsible for keeping a pipeline process alive and monitoring its status.
+    It also handles the streaming of input and output frames to the pipeline.
+    """
+    def __init__(
+        self,
+        pipeline: str,
+        params: dict,
+    ):
+        self.pipeline = pipeline
+        self.params = params
+        self.monitoring_callback = _noop_callback
+
+        self.process = None
+        self.monitor_task = None
+        self.status = PipelineStatus(
+            pipeline=pipeline, start_time=time.time()
+        ).update_params(params, False)
+
+    async def start(self):
+        # Start the pipeline process and initialize timing
+        self.process = PipelineProcess.start(self.pipeline, self.params)
+        if self.process is None:
+            raise RuntimeError("Failed to start PipelineProcess")
+        # Launch the monitor loop as a background task
+        self.monitor_task = asyncio.create_task(self._monitor_loop())
+
+    async def stop(self):
+        if self.monitor_task:
+            self.monitor_task.cancel()
+            try:
+                await self.monitor_task
+            except asyncio.CancelledError:
+                pass
+            self.monitor_task = None
+        if self.process:
+            await self.process.stop()
+            self.process = None
+
+    async def reset_stream(
+        self,
+        request_id: str,
+        stream_id: str,
+        params: dict,
+        monitoring_callback: Callable[[dict], Awaitable[None]],
+    ):
+        if not self.process:
+            raise RuntimeError("Process not running")
+        self.status = PipelineStatus(
+            pipeline=self.pipeline, start_time=self.status.start_time
+        )
+        self.monitoring_callback = monitoring_callback
+        self.process.reset_stream(request_id, stream_id)
+        await self.update_params(params)
+
+    def send_input(self, frame: InputFrame):
```

Comment:
what's the difference between ProcessGuardian handling input/output of frames vs the PipelineStreamer? I'm having a bit of confusion coming up with a mental model